components/server/src/workspace/workspace-starter.ts

```diff
@@ -530,54 +530,86 @@ export class WorkspaceStarter {
                 ...user,
             };
 
-            // choose a cluster and start the instance
-            let resp: StartWorkspaceResponse.AsObject | undefined = undefined;
-            let retries = 0;
-            try {
-                for (; retries < MAX_INSTANCE_START_RETRIES; retries++) {
-                    resp = await this.tryStartOnCluster({ span }, startRequest, euser, workspace, instance);
-                    if (resp) {
-                        break;
+            const ideUrlPromise = new Deferred<string>();
+
+            (async () => {
+                try {
+                    // choose a cluster and start the instance
+                    let resp: StartWorkspaceResponse.AsObject | undefined = undefined;
+                    let retries = 0;
+                    try {
+                        for (; retries < MAX_INSTANCE_START_RETRIES; retries++) {
+                            resp = await this.tryStartOnCluster({ span }, startRequest, euser, workspace, instance);
+                            if (resp) {
+                                break;
+                            }
+                            await new Promise((resolve) =>
+                                setTimeout(resolve, INSTANCE_START_RETRY_INTERVAL_SECONDS * 1000),
+                            );
+                        }
+                    } catch (err) {
+                        await this.failInstanceStart({ span }, err, workspace, instance);
+                        throw new StartInstanceError("startOnClusterFailed", err);
+                    }
+
+                    if (!resp) {
+                        const err = new Error("cannot start a workspace because no workspace clusters are available");
+                        await this.failInstanceStart({ span }, err, workspace, instance);
+                        throw new StartInstanceError("clusterSelectionFailed", err);
                     }
-                    await new Promise((resolve) => setTimeout(resolve, INSTANCE_START_RETRY_INTERVAL_SECONDS * 1000));
+                    increaseSuccessfulInstanceStartCounter(retries);
+
+                    span.log({ resp: resp });
+                    ideUrlPromise.resolve(resp.url);
+                } catch (err) {
+                    ideUrlPromise.reject(err);
                 }
-            } catch (err) {
-                await this.failInstanceStart({ span }, err, workspace, instance);
-                throw new StartInstanceError("startOnClusterFailed", err);
-            }
+            })();
 
-            if (!resp) {
-                const err = new Error("cannot start a workspace because no workspace clusters are available");
-                await this.failInstanceStart({ span }, err, workspace, instance);
-                throw new StartInstanceError("clusterSelectionFailed", err);
-            }
-            increaseSuccessfulInstanceStartCounter(retries);
+            const intervalHandle = setInterval(async () => {
```

Comment:
This new code is started in parallel and polls the DB every 50 ms.

Reply:
This should be `repeat` instead of `setInterval` to avoid chaining issues/races.

Reply:
What if we notice in production that we break implicit assumptions made by clients when we return here? :thinking:
Do we revert + deploy again?

Reply:
As it stands we would need to do that. Milan suggested adding a feature flag so we can act faster.